diag_manager/fms_diag_buffer.F90

```diff
@@ -0,0 +1,1097 @@
+  contains
+  procedure :: allocate_buffer => allocate_buffer_2d
+  procedure :: get_buffer_data => get_buffer_2d
+  procedure :: initialize_buffer => initialize_buffer_2d
+  procedure :: add_to_buffer => add_to_buffer_2d
+end type buffer2d
+
+!> 3D buffer type to extend fmsDiagBuffer_class
+type, extends(fmsDiagBuffer_class) :: buffer3d
+  class(*), allocatable :: buffer(:,:,:) !< 3D numeric data array
+  contains
+  procedure :: allocate_buffer => allocate_buffer_3d
+  procedure :: get_buffer_data => get_buffer_3d
+  procedure :: initialize_buffer => initialize_buffer_3d
+  procedure :: add_to_buffer => add_to_buffer_3d
+end type buffer3d
+
+!> 4D buffer type to extend fmsDiagBuffer_class
+type, extends(fmsDiagBuffer_class) :: buffer4d
+  class(*), allocatable :: buffer(:,:,:,:) !< 4D numeric data array
+  contains
+  procedure :: allocate_buffer => allocate_buffer_4d
+  procedure :: get_buffer_data => get_buffer_4d
+  procedure :: initialize_buffer => initialize_buffer_4d
+  procedure :: add_to_buffer => add_to_buffer_4d
+end type buffer4d
+
+!> 5D buffer type to extend fmsDiagBuffer_class
+type, extends(fmsDiagBuffer_class) :: buffer5d
+  class(*), allocatable :: buffer(:,:,:,:,:) !< 5D numeric data array
+  contains
+  procedure :: allocate_buffer => allocate_buffer_5d
+  procedure :: get_buffer_data => get_buffer_5d
+  procedure :: initialize_buffer => initialize_buffer_5d
+  procedure :: add_to_buffer => add_to_buffer_5d
+end type buffer5d
+
+! public types
+public :: buffer0d
+public :: buffer1d
+public :: buffer2d
+public :: buffer3d
+public :: buffer4d
+public :: buffer5d
+public :: fmsDiagBuffer_class
+public :: fmsDiagBufferContainer_type
+! public routines
+public :: fms_diag_buffer_init 
+! Module variables
+logical,private :: module_is_initialized = .false. !< Flag indicating if the module is initialized
+integer, parameter :: DEFAULT_BUFFER_LIST_SIZE = 64
+
+
+logical, parameter, private :: DEBUG = .true. !< debugging output
+
+! TODO interface with function declarations needed for deferred routines
+!abstract interface
+  !integer function allocate_buffer(buffobj, mold) &
+  !result(rslt)
+    !import buffer0d 
+    !!import time_type 
+    !class(fmsDiag), intent(inout) :: buffobj !< scalar buffer object
+    !class(*),intent(in) :: mold !< allocates to the type of mold
+    !!type(time_type), intent(in), optional :: init_time
+  !end function
+!end interface
+!
+contains
+
+
+!!--------module routines
+
+!> Initializes a list of diag buffers
+!! and sets the buffer_id's to DIAG_NOT_REGISTERED
+logical function fms_diag_buffer_init(buffobjs, buff_dims)
+  type(fmsDiagBufferContainer_type), allocatable, intent(out) :: buffobjs(:) !< an array of buffer container types to allocate
+  integer, intent(in)                                          :: buff_dims !< number of dimensions needed for the buffer data
+  integer :: i
+
+  allocate(buffobjs(DEFAULT_BUFFER_LIST_SIZE))
```

Comment:
This has to be the size of the diag_yaml%diag_fields

Reply:
`diag_yaml%diag_fields` can be passed in from `fms_diag_object%init`.  That's the number of fields for all files, right?

Reply:
Yeah